### DISCUSSION TITLE
{**d1, **d2} vs dict(**d1, **d2)

### DISCUSSION THREAD

**Discussion by Beormund** [micropython/micropython#10058] [Using MicroPython] [answered]
Is this intended behaviour:
```py
# MicroPython v1.19.1-469-g8a1c9464e on 2022-10-04; Raspberry Pi Pico W with RP2040
d1 = {'a': 1}
d2 = {'b': 2}

# This works:
dict(**d1, **d2)
# >>> {'a': 1, 'b': 2}

# This fails:
{**d1, **d2}
# >>> SyntaxError: invalid syntax
```
No problem at all to use the constructor - but just curious about the latter.

**Comment by rkompass** [comment]
In my Python 3.10.8 console `{**d1, **d2}` works, so it should not be intended behaviour, as MP tries to be compatible.

**Comment by jimmo** [comment, marked as answer]
`{**d1, **d2}` is a Python 3.5 feature, described in [PEP 448](https://peps.python.org/pep-0448/). MicroPython is Python 3.4 with a selection of features from later versions (including some features from that same PEP).

This is a slightly controversial feature, to the extent that they "replaced" it with the | operator -- i.e. `d1 | d2` -- described in [PEP 584](https://peps.python.org/pep-0584/) (which was added in Python 3.9)

**Comment by peterhinch** [comment]
https://github.com/micropython/micropython/pull/5807 seems to indicate that PEP584 has only been implemented for function args.

**Reply by dlech** [reply]
I think you mean PEP448, but yes, that is correct.

**Comment by jimmo** [comment]
FYI: PEP 584 is now supported https://github.com/micropython/micropython/pull/8523

**Reply by peterhinch** [reply]
@jimmo I'm puzzled by the latest build, which I've compared with CPython V3.11.0. The PEP584 syntax does indeed work, which is great. Under CPython the following PEP448 syntax works, but not under MP:
```python
d = {**a, **b}
```
So your statement
> they "replaced" it with the | operator -- i.e. d1 | d2

seems wrong. As usual the CPython maintainers kept compatibility with existing code. Should MP do likewise?

The following shows a difference with CPython. It works fine if the dicts have no common keys:
```python
def foo(**x):
    print(x)
foo(**a, **b)  # a and b are dicts
```
With these dicts, behaviour differs:
```python
a = {'one': 1, 'two': 2}
z = {'one': 1, 'five': 5}
```
MP:
```python
>>> foo(**a, **z)
{'two': 2, 'one': 1, 'five': 5}
```
CPython:
```python
>>> foo(**a, **z)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __main__.foo() got multiple values for keyword argument 'one'
```

**Reply by dlech** [reply]
Nice find. I started #10083 to discuss what we could do about the CPython difference (it seems like an independent discussion from the `|` operator).

**Reply by peterhinch** [reply]
I have raised https://github.com/micropython/micropython/issues/10091 to point out the failure of PEP448 syntax.

**Reply by jimmo** [reply]
@peterhinch 

> So your statement `they "replaced" it with the | operator -- i.e. d1 | d2` seems wrong

Yes, sorry I did phrase that quite poorly. What I meant to say was that the `d1 | d2` replaces `{**d1, **d2}` as the easiest way to union two dictionaries. See the ["Motivation" section of PEP584](https://peps.python.org/pep-0584/#motivation) for the quote from Guido.

It might be a bit strange if MicroPython never ends up supporting the PEP448 style... I don't think it's something that has come up much before this thread (even though it's a 3.5 feature). Thanks for raising #10091.
